### ISSUE TITLE
Azure ARM - Windows nil pointer exception

### ISSUE THREAD

**Issue by tombuildsstuff** [mitchellh/packer#3581]
I'm using Packer v0.10.1 installed via Brew on OS X 10.11.5 to build one of the [Azure Windows Templates](https://github.com/mitchellh/packer/blob/master/examples/azure/windows.json) located in the master branch

When running this - I'm see a nil-pointer exception:

https://gist.githubusercontent.com/tombuildsstuff/dbf909d44761f352aaeb4f7a88cdc53c/raw/5764667b6fa044cef7bed7ae0342118e3dd03338/gistfile1.txt

**Comment by cbednarski** [comment]
Thanks for the report! I think there is a fix for this bug pending or in master; it should be fixed in the next release. There is some additional context here: https://github.com/mitchellh/packer/issues/3456

Will wait for @boumenot to verify before closing.

**Comment by boumenot** [comment]
I think this is a different issue.  The builder paniced while trying to dereference a pointer from the SDK's response.  The code is trying to get the URL of a certificate that was previously uploaded to key vault, but it appears that it did not succeed.  I have not seen this before.  

I have seen cases where the keyvault is not ready immediately, and the code handles that by polling.  It stops polling once the SDK returns success.  It appears the SDK returned success, but still did not contain the requisite certificate.  Gah!  I can check for nil, and force it to poll again.  From the gist you included, I can see that the builder never polled.  (Almost always it succeeds immediately, so one ever sees poll messages.)

@tombuildsstuff, how reproducible is this error?  If it happens frequently, would you please re-run with [PACKER_DEBUG=1](https://www.packer.io/docs/other/debugging.html).

I'll prepare a fix.

**Comment by tombuildsstuff** [comment]
@boumenot @cbednarski thanks for getting back to me about this. This error is reproducible every time - I've done a [debug trace](https://gist.github.com/tombuildsstuff/7ec251a1c74ce24ad540bc46b3ce1159) as requested - although I've sanitised a few fields - let me know if I've sanitised too much.

Thanks!

**Comment by boumenot** [comment]
Thanks, this helps out immensely.  This is similar to another [issue](https://github.com/Azure/azure-sdk-for-go/issues/334) that I've logged against the Azure SDK.

The error message indicates that the packer user does not have access to the certificate in the KeyVault.  My **first** guess is that your ObjectID is incorrect.  The Azure builds under an SPN, and the KeyVault is protected using the ObjectID of the SPN.  The builder uses KeyVault as a middle man for injecting a certificate in the VM to authenticate WinRM communication.

You can check the ObjectID using Azure XPlat CLI.

```batch
azure ad sp show --spn <client_id>
```

There are two action items for me on this issue.

1. Remove the need for the user to enter an ObjectID.  I should be able to look it up using existing information.  There are already too many GUIDs IMHO.
1. Better guard against this issue in the code.  Hopefully the SDK team can deliver a fix, but I can mitigate it fairly easily.

**Comment by boumenot** [comment]
@paulmey is looking into 1.

The Azure/azure-sdk-for-go [does not yet have support][1] for the KeyVault.  (I should have known this because I imported the code in the first place.)  I've updated the code to better handle these types of errors, and hopefully provide a more informed message.  The real fix is (1), which is being worked now.  The less opportunity for mistakes the better.

[1]: https://github.com/Azure/azure-rest-api-specs/issues/67
